pkg/controllers/namespace/subnet_handler.go

```diff
@@ -0,0 +1,336 @@
+	servicecommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+)
+
+// extractSubnetPath extracts the subnet name, VPC name, and project name from a subnet path
+func (r *NamespaceReconciler) extractSubnetPath(sharedSubnetPath string) (subnetName, projectName, vpcName string, err error) {
+	// Format: /orgs/default/projects/proj-1/vpcs/vpc-1/subnets/subnet-1
+	re := regexp.MustCompile(`/projects/([^/]+)/vpcs/([^/]+)/subnets/([^/]+)`)
+	matches := re.FindStringSubmatch(sharedSubnetPath)
+	if len(matches) < 4 {
+		return "", "", "", fmt.Errorf("invalid subnet path format: %s", sharedSubnetPath)
+	}
+
+	projectName = matches[1]
+	vpcName = matches[2]
+	subnetName = matches[3]
+	// "_" is not allowed in the name, so replace it with "-"
+	subnetName = strings.ReplaceAll(subnetName, "_", "-")
+	return subnetName, projectName, vpcName, nil
+}
+
+// convertSubnetPathToAssociatedResource converts a subnet path to the associated resource format
+// e.g., /orgs/default/projects/proj-1/vpcs/vpc-1/subnets/subnet-1 -> proj-1:vpc-1:subnet-1
+func (r *NamespaceReconciler) convertSubnetPathToAssociatedResource(sharedSubnetPath string) (string, error) {
+	// Extract the subnet name, VPC name and project name using extractSubnetPath
+	subnetName, projectName, vpcName, err := r.extractSubnetPath(sharedSubnetPath)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s:%s:%s", projectName, vpcName, subnetName), nil
+}
+
+// getNSXSubnet retrieves the NSX subnet from the NSX API
+func (r *NamespaceReconciler) getNSXSubnet(ns string, sharedSubnetPath string) (*model.VpcSubnet, error) {
+	vpcInfoList := r.VPCService.ListVPCInfo(ns)
+	if len(vpcInfoList) == 0 {
+		return nil, fmt.Errorf("no VPC info found for namespace: %s", ns)
+	}
+
+	vpcInfo := vpcInfoList[0]
+	nsxSubnets, err := r.CommonService.NSXClient.SubnetsClient.List(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, nil, nil, nil, nil, nil, nil)
+	if err != nil {
+		log.Error(err, "Failed to list nsx subnets")
+		return nil, err
+	}
+
+	for _, subnet := range nsxSubnets.Results {
+		if *subnet.Path == sharedSubnetPath {
+			return &subnet, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// buildSubnetCR creates a Subnet CR object with the given parameters
+func (r *NamespaceReconciler) buildSubnetCR(ns, subnetName, vpcFullName, associatedName string, nsxSubnet *model.VpcSubnet) *v1alpha1.Subnet {
+	// Create the Subnet CR
+	subnetCR := &v1alpha1.Subnet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      subnetName,
+			Namespace: ns,
+			Annotations: map[string]string{
+				servicecommon.AnnotationAssociatedResource: associatedName,
+			},
+		},
+		Spec: v1alpha1.SubnetSpec{
+			VPCName:             vpcFullName,
+			EnableVLANExtension: true,
+		},
+	}
+
+	// Initialize subnetCR from nsxSubnet if available
+	if nsxSubnet != nil {
+		r.mapNSXSubnetToSubnetCR(subnetCR, nsxSubnet)
+	} else {
+		// Use default values if nsxSubnet is not available
```

Comment:
If there is no NSX subnet, we do not need create Subnet CR for it.

Reply:
fixed